src/messaging/ExchangeMgr.cpp

```diff
@@ -230,10 +232,11 @@ void ExchangeManager::OnMessageReceived(const PacketHeader & packetHeader, const
                         packetHeader.GetDestinationGroupId().Value());
     }
 
+    // Do not handle unsolicited messages on a inactive session.
     // If it's not a duplicate message, search for an unsolicited message handler if it is marked as being sent by an initiator.
     // Since we didn't find an existing exchange that matches the message, it must be an unsolicited message. However all
     // unsolicited messages must be marked as being from an initiator.
-    if (!msgFlags.Has(MessageFlagValues::kDuplicateMessage) && payloadHeader.IsInitiator())
+    if (session->IsActiveSession() && !msgFlags.Has(MessageFlagValues::kDuplicateMessage) && payloadHeader.IsInitiator())
```

Comment:
Shouldn't this be a dedicated check for `IsActiveSession` and return early if that is not true? Doing it this way means that we'll still proceed further if the message need an ACK to be sent right?

Reply:
Here the session may always be active, but let's consider it logically.

Even the session is marked as pending for removal, on the other hand, the peer doesn't have the information, so there can be messages being arrived over this session, we shall process these messages, and respond with proper reply and ack. But shall not create new exchanges over the session.

So the check is fine here.